File: tests/analytical/sphere.py
#! /usr/bin/env python

import numpy as np
import sdf_tools

n = 5

offs = (0., 0., 0.)
dims = (n, n, n)
exts = (1.0, 1.0, 1.0)
center = (0.5, 0.5, 0.5)
radius = 0.4

grid = sdf_tools.Grid.Uniform(dims, offs, exts)

sphere = sdf_tools.Sdf.Sphere(center, radius, inside=True)

sphere.apply(grid)

np.savetxt("grid.txt", grid.get())

# TEST: analytical.sphere
# cd analytical
# rm -rf *.txt
# ./sphere.py
# mv grid.txt grid.out.txt

File: src/sdf_tools/core/sdf/sphere.cpp
#include "sphere.h"

#include <sdf_tools/core/utils/helper_math.h>

namespace sdf_tools {
namespace sdf {

SdfSphere::SdfSphere(real3 center, real radius, bool inside) :
    center_(center),
    radius_(radius),
    sign_(inside ? 1 : -1)
{}

real SdfSphere::at(real3 r) const
{
    return sign_ * (length(r - center_) - radius_);
}

} // namespace sdf
} // namespace sdf_tools
